---
title: "Adelaide Ticks Part 2"
author: "Kelly Pierce"
date: "August 1, 2017"
output: html_document
---

#Preprocessing

Most recently generated VCF file is 

`/Users/Kelly/Desktop/Flinders_RADseq/dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes.vcf`

Command lines for processing the genotype data:

*Representative filtering commands* (not inclusive of all commands executed)

`$ vcftools --vcf dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes.vcf --min-meanDP 10 --minGQ 25 --maf 0.05 --max-missing 0.75 --recode --out dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_maf05_missing75`

VCFtools output:

  After filtering, kept 422 out of 422 Individuals
  Outputting VCF file...
  After filtering, kept 263 out of a possible 565560 Sites
  
**Minor allele frequency filter may not make sense, as this VCF file has duplicates (leg/body) and larval pools**
  
`$ vcftools --vcf dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes.vcf --min-meanDP 10 --minGQ 25 --max-missing 0.75 --recode --out dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing75`

VCFtools output

After filtering, kept 422 out of 422 Individuals
Outputting VCF file...
After filtering, kept 1007 out of a possible 565560 Sites
Run Time = 52.00 seconds

`$ sed '/^##/ d' dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing75.recode.vcf | awk '!array[$1]++' > dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing75_firstSNP.vcf`

This file has 257 lines.

*Get the 012 file*
`$ vcftools --vcf dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing75_firstSNP.vcf --012 --out dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing75_firstSNP_012`

Files transfered to macbook...

*Convert the 012 file for input into adegenet*
`kellys-mbp-2:Flinders_RADseq Kelly$ python ~/Desktop/CSU_ChronicWasting/Scripts/VCF_to_Adegenet.py -o dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing75_firstSNP_012.012 -n 422 -a dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing75_firstSNP_012.adegenet.snp`

#Adegenet analysis

```{r dataImportExplore}
require(adegenet)
setwd('~/Desktop/Flinders_RADseq/')
tick<-read.snp('dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing75_firstSNP_012.adegenet.snp')

# data exploration
class(tick)
names(tick)
print(object.size(tick$gen), unit='Mb')
class(tick$gen)
class(tick$other)
length(tick$other)

# to look at the actual genotypes in a single snpbin object:
head(as.integer(tick@gen[[1]]))

glPlot(tick)
hist(glMean(tick))
```

##PCA

```{r PCA.1}
# first pass
pca1 <- glPca(tick)
dim(pca1$scores)
scatter(pca1)
plot(pca1$scores[,1], pca1$scores[,2])
```

Do clusters correspond to any aspects of study design?

```{r PCA.2}
# get library & tissue info
metaData <- read.table('~/Desktop/Flinders_RADseq/barcodes_tissues_libraries.txt')
names(metaData) <- c('ID', 'barcode', 'enzyme', 'tissue', 'libName')
# reconstruct the file name that will be found in the adegenet objects
# this includes Is in the place of some 1s due to an earlier error
bodies_only <- metaData[which(metaData$tissue == 'bodies'),]
bodies_only$reconstructedID <- paste('/mnt/HGST4TB/legs_larvae_reference_assembly_output/dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref//DBR_filtered_sequences_', paste(bodies_only$ID, bodies_only$barcode, 'psti.R1_bodies.sorted.bam', sep='_'), sep='')
legs_larvae_only <- metaData[which(metaData$tissue != 'bodies'),]
legs_larvae_only$reconstructedID <- paste('/mnt/HGST4TB/legs_larvae_reference_assembly_output/dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref//DBR_filtered_sequences_', paste(legs_larvae_only$ID, legs_larvae_only$barcode, 'psti.R1_legs_larvae.sorted.bam', sep='_'), sep='')
metaData_final <- rbind(bodies_only, legs_larvae_only)
# now fix the I vs 1 issue
metaData_final$fixedID <- gsub('I', '1', metaData_final$ID)

lib <- c()
tissue <- c()

for(t in 1:length(row.names(pca1$scores))){
  m <- metaData_final[which(as.character(metaData_final$reconstructedID) == as.character(row.names(pca1$scores)[t])),]
  #print(m)
  lib <- c(lib, as.character(m$libName))
  tissue <- c(tissue, as.character(m$tissue))
}

tissue.colors <- c("#fa9fb5", "#c51b8a", "#7a0177")
library.colors <- c('#f7fcfd','#e0ecf4','#bfd3e6','#9ebcda','#8c96c6','#8c6bb1','#88419d','#6e016b')

tc_column <- c()
lc_column <- c()
for(t in 1:length(metaData_final[,1])){
  tissue.type <- metaData_final[t,]$tissue
  tc <- tissue.colors[as.factor(tissue.type)]
  tc_column <- c(tc_column, tc)
  #print(tissue.type)
  #print(tc)
  lib.type <- metaData_final[t,]$libName
  lc <- library.colors[as.factor(lib.type)]
  lc_column <- c(lc_column, lc)
}
metaData_final$tissue_color <- tc_column
metaData_final$lib_color <- lc_column

pca_df <- data.frame(pca1$scores)
#pca_df$names <- rownames(pca_df)
library(data.table)
pca_df <- data.frame(setDT(pca_df, keep.rownames = TRUE)[])
pca_df$rn <- as.character(pca_df$rn)
pca_df_merged <- merge(pca_df, metaData_final, by.x = "rn", by.y = "reconstructedID", all.x = TRUE)


plot(pca_df_merged$PC1, pca_df_merged$PC2, col = pca_df_merged$tissue_color, pch=16)
legend('bottomleft', legend = levels(pca_df_merged$tissue), 
       col = tissue.colors, pch = 16, bty = 'n')

plot(pca_df_merged$PC1, pca_df_merged$PC2, col = pca_df_merged$lib_color, pch=16)


# no clear correspondance to tissue or library

xmin = min(pca_df_merged$PC1)
xmax = max(pca_df_merged$PC1)
ymin = min(pca_df_merged$PC2)
ymax = max(pca_df_merged$PC2)
pdf('~/Desktop/Flinders_RADseq/Tick_RADseq_Replicate_PCA.pdf', height = 7, width = 7)
finished_list <- c()
for(id in pca_df_merged$fixedID){
  #print(id)
  rm(duplicates)
  duplicates <- pca_df_merged[which(pca_df_merged$fixedID == id),]
  if(dim(duplicates)[1] >= 2){
    if(!(id %in% finished_list)){
      plot(pca_df_merged$PC1, pca_df_merged$PC2, col = 'gray', 
         main = paste('Focal Sample: ', id, ' (N=', length(duplicates$rn), ")", sep = ''),
         xlim = c(xmin, xmax), ylim = c(ymin, ymax), pch = 16,
         xlab = 'PC1', ylab = 'PC2', las = 1)
    points(duplicates$PC1, duplicates$PC2, col = duplicates$tissue_color, pch = 20)
    points(duplicates$PC1, duplicates$PC2, col = 'black')
    legend('bottomleft', legend = c(levels(pca_df_merged$tissue), 'others'), 
       col = c(tissue.colors, 'gray'), pch = 20, bty = 'n')
    finished_list <- c(finished_list, id)
    }
  }
}
dev.off()
```

The technical replicates don't have the same PC values. I suspect that the fraction of missing data allowed is responsible for this. I'm going to filter at different "maxmissing" thresholds to see if that changes the result.

-- 85% --

`$ vcftools --vcf dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes.vcf --min-meanDP 10 --minGQ 25 --max-missing 0.85 --recode --out dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing85`

`$ sed '/^##/ d' dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing85.recode.vcf | awk '!array[$1]++' > dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing85_firstSNP.vcf`

`$ vcftools --vcf dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing85_firstSNP.vcf --012 --out dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing85_firstSNP_012`

Files transfered to macbook...
`kellys-mbp-2:Flinders_RADseq Kelly$ python ~/Desktop/CSU_ChronicWasting/Scripts/VCF_to_Adegenet.py -o dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing85_firstSNP_012.012 -n 422 -a dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing85_firstSNP_012.adegenet.snp`

```{r functions}
reconstruct_IDs <- function(meta.data, 
                            sample.name.prefix,
                            sample.name.suffix){
  # reconstruct the file name that will be found in the adegenet objects
  # this includes Is in the place of some 1s due to an earlier error
  meta.data$reconstructedID <- paste(sample.name.prefix, 
                                       paste(meta.data$ID,
                                             meta.data$barcode,
                                             sample.name.suffix,
                                             sep='_'), sep='')
  meta.data$fixedID <- gsub('I', '1', meta.data$ID)
  return(meta.data)
}

merge.pca.meta.data <- function(pca.data,
                                meta.data){
  require(data.table)
  pca.df <- data.frame(pca.data$scores)
  pca.df <- data.frame(setDT(pca.df, keep.rownames = TRUE)[])
  pca.df$rn <- as.character(pca.df$rn)
  pca.df.merged <- merge(pca.df, meta.data, by.x = "rn", by.y = "reconstructedID", all.x = TRUE)
  return(pca.df.merged)
}

assign.colors <- function(full.data,
                          variable.column,
                          color.vector,
                          color.column.name){
  color.column <- c()
  
  for(var in 1:length(variable.column)){
    var.type <- variable.column[var]
    var.color <- color.vector[as.factor(var.type)]
    color.column <- c(color.column, var.color)
  }
  full.data$new_column <- color.column
  names(full.data)[length(names(full.data))] <- color.column.name
  return(full.data)
}

parse_legend <- function(factor.data, color.data){
  legend.text <- c()
  fill.color <- c()
  if(length(factor.data) != length(color.data)){
    return('Data series of unequal lengths.')
  }else{
    for(i in 1:length(factor.data)){
      
    }
  }
  
}

replicates_pca_plot <- function(full.data,
                                output.file,
                                variable.column.name,
                                color.column.name){
  xmin = min(full.data$PC1)
  xmax = max(full.data$PC1)
  ymin = min(full.data$PC2)
  ymax = max(full.data$PC2)
  pdf(output.file, height = 7, width = 7)
  dt <- table(full.data$fixedID)
  dtn <- names(dt[which(dt > 1)])  
  #finished_list <- c()
  for(id in dtn){
    duplicates <- full.data[which(full.data$fixedID == id),]
    print(duplicates)
    vcn <- quote(variable.column.name)
    ccn <- quote(color.column.name)
    #duplicates[,eval(vcn)] <- factor(duplicates[,eval(vcn)])
    #duplicates[,eval(ccn)] <- factor(as.factor(as.character(duplicates[,eval(ccn)])))
    plot(full.data$PC1, full.data$PC2, col = 'gray', 
             main = paste('Focal Sample: ', id, ' (N=', length(duplicates$rn), ")", 
                        sep = ''),
             xlim = c(xmin, xmax), ylim = c(ymin, ymax), pch = 16,
             xlab = 'PC1', ylab = 'PC2', las = 1)
    points(duplicates$PC1, duplicates$PC2, col = duplicates[,eval(ccn)], pch = 20)
    points(duplicates$PC1, duplicates$PC2, col = 'black')
    legend.text <- c()
    legend.color <- c()
    for(d in 1:length(duplicates[,1])){
      d.vcn <- as.character(duplicates[d,eval(vcn)])
      d.ccn <- as.character(duplicates[d,eval(ccn)])
      if(!(d.vcn %in% legend.text)){
        legend.text <- c(legend.text, d.vcn)
      }
      if(!(d.ccn %in% legend.color)){
        legend.color <- c(legend.color, d.ccn)
      }
    }
    legend('bottomleft', 
               legend = c(legend.text,'others'),
               col = c(legend.color, 'gray'), 
               pch = 20, bty = 'n')
  }
  dev.off()
}
```

```{r function_calls_85}

setwd('~/Desktop/Flinders_RADseq/minDP10_minGQ25_missing85/')
tick.85<-read.snp('dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing85_firstSNP_012.adegenet.snp')
pca.85 <- glPca(tick.85)
dim(pca.85$scores)
plot(pca.85$scores[,1], pca.85$scores[,2])

metaData <- read.table('~/Desktop/Flinders_RADseq/barcodes_tissues_libraries.txt')
names(metaData) <- c('ID', 'barcode', 'enzyme', 'tissue', 'libName')
bodies_only <- metaData[which(metaData$tissue == 'bodies'),]
legs_larvae_only <- metaData[which(metaData$tissue != 'bodies'),]

sn.prefix = '/mnt/HGST4TB/legs_larvae_reference_assembly_output/dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref//DBR_filtered_sequences_'
sn.bodies.suffix = 'psti.R1_bodies.sorted.bam'
sn.legs_larvae.suffix = 'psti.R1_legs_larvae.sorted.bam'

bodies.reconstructed <- reconstruct_IDs(meta.data = bodies_only,
                                        sample.name.prefix = sn.prefix,
                                        sample.name.suffix = sn.bodies.suffix)
legs_larvae.reconstructed <- reconstruct_IDs(meta.data = legs_larvae_only,
                                        sample.name.prefix = sn.prefix,
                                        sample.name.suffix = sn.legs_larvae.suffix)

meta.data.final <- rbind(bodies.reconstructed, legs_larvae.reconstructed)
merged.data.85 <- merge.pca.meta.data(pca.85, meta.data.final)
  
tissue.colors <- c("#fa9fb5", "#c51b8a", "#7a0177")
library.colors <- c('#f7fcfd','#e0ecf4','#bfd3e6',
                    '#9ebcda','#8c96c6','#8c6bb1',
                    '#88419d','#6e016b')
full.data.colors <- assign.colors(merged.data.85, merged.data.85$tissue, 
                                  tissue.colors, 'tissue.color')
full.data.colors.85 <- assign.colors(full.data.colors, full.data.colors$libName,
                                     library.colors, 'library.color')

replicates_pca_plot(full.data = full.data.colors.85,
                    output.file = "~/Desktop/Flinders_RADseq/Tick_RADseq_Replicate_PCA_maxmissing85.pdf",
                    variable.column = 'tissue',
                    color.column = 'tissue.color')

replicates_pca_plot(full.data = full.data.colors.85,
                    output.file = "~/Desktop/Flinders_RADseq/Tick_RADseq_Replicate_PCA_maxmissing85_byLib.pdf",
                    variable.column.name <- 'libName',
                    color.column.name = 'library.color')

```

an idea for a formal test -- calculate pairwise euclidian distances between all the samples and ask if the distance for replicates is closer than expected by random chance...

-- 90% --
(95% threshold left only 1 site...)

`$ vcftools --vcf dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes.vcf --min-meanDP 10 --minGQ 25 --max-missing 0.90 --recode --out dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing90`

`$ sed '/^##/ d' dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing90.recode.vcf | awk '!array[$1]++' > dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing90_firstSNP.vcf`

`$ vcftools --vcf dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing90_firstSNP.vcf --012 --out dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing90_firstSNP_012`

Files transfered to macbook...
`kellys-mbp-2:Flinders_RADseq Kelly$ python ~/Desktop/CSU_ChronicWasting/Scripts/VCF_to_Adegenet.py -o dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing90_firstSNP_012.012 -n 422 -a dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing90_firstSNP_012.adegenet.snp`

```{r function_calls_90}

setwd('~/Desktop/Flinders_RADseq/minDP10_minGQ25_missing90/')
tick.90<-read.snp('dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing90_firstSNP_012.adegenet.snp')
pca.90 <- glPca(tick.90)
dim(pca.90$scores)
plot(pca.90$scores[,1], pca.90$scores[,2])

merged.data.90 <- merge.pca.meta.data(pca.90, meta.data.final)
  
tissue.colors <- c("#fa9fb5", "#c51b8a", "#7a0177")
library.colors <- c('#f7fcfd','#e0ecf4','#bfd3e6',
                    '#9ebcda','#8c96c6','#8c6bb1',
                    '#88419d','#6e016b')
full.data.colors <- assign.colors(merged.data.90, merged.data.90$tissue, 
                                  tissue.colors, 'tissue.color')
full.data.colors.90 <- assign.colors(full.data.colors, full.data.colors$libName,
                                     library.colors, 'library.color')

replicates_pca_plot(full.data = full.data.colors.90,
                    output.file = "~/Desktop/Flinders_RADseq/Tick_RADseq_Replicate_PCA_maxmissing90.pdf",
                    variable.column = full.data.colors.90$tissue,
                    color.vector = tissue.colors)

```

-- 90% + minor allele frequency filtering

`$ vcftools --vcf dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes.vcf --min-meanDP 10 --minGQ 25 --maf 0.01 --max-missing 0.90 --recode --out dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_maf01_missing90`

`$ sed '/^##/ d' dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_maf01_missing90.recode.vcf | awk '!array[$1]++' > dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_maf01_missing90_firstSNP.vcf`

`$ vcftools --vcf dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_maf01_missing90_firstSNP.vcf --012 --out dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_maf01_missing90_firstSNP_012`

Files transfered to macbook...
`kellys-mbp-2:Flinders_RADseq Kelly$ python ~/Desktop/CSU_ChronicWasting/Scripts/VCF_to_Adegenet.py -o dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_maf01_missing90_firstSNP_012.012 -n 422 -a dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_maf01_missing90_firstSNP_012.adegenet.snp`

```{r function_calls_90_maf01}
setwd('~/Desktop/Flinders_RADseq/minDP10_minGQ25_maf01_missing90/')
tick.90.01<-read.snp('dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_maf01_missing90_firstSNP_012.adegenet.snp')
pca.90.01 <- glPca(tick.90.01)
dim(pca.90.01$scores)
plot(pca.90.01$scores[,1], pca.90.01$scores[,2])

merged.data.90.01 <- merge.pca.meta.data(pca.90.01, meta.data.final)
  
tissue.colors <- c("#fa9fb5", "#c51b8a", "#7a0177")
library.colors <- c('#f7fcfd','#e0ecf4','#bfd3e6',
                    '#9ebcda','#8c96c6','#8c6bb1',
                    '#88419d','#6e016b')
full.data.colors.90.01 <- assign.colors(merged.data.90.01, merged.data.90.01$tissue, 
                                  tissue.colors, 'tissue.color')
full.data.colors.90.01 <- assign.colors(full.data.colors, full.data.colors$libName,
                                     library.colors, 'library.color')

replicates_pca_plot(full.data = full.data.colors.90.01,
                    output.file = "~/Desktop/Flinders_RADseq/Tick_RADseq_Replicate_PCA_maxmissing90_maf01.pdf",
                    variable.column = full.data.colors.90.01$tissue,
                    color.vector = tissue.colors)

replicates_pca_plot(full.data = full.data.colors.90.01,
                    output.file = "~/Desktop/Flinders_RADseq/Tick_RADseq_Replicate_PCA_maxmissing90_maf01_byLib.pdf",
                    variable.column = full.data.colors.90.01$libName,
                    color.vector = library.colors)

```

Does tissue type matter?
```{r}
distance <- function(x1, y1, x2, y2){
  return(sqrt(((x2-x1)**2) + ((y2-y1)**2)))
}

replicates_by_tissue_library <- function(full.data, output.file){
  dt <- table(full.data$fixedID)
  dtn <- names(dt[which(dt > 1)])  
  reps.by.tissue.lib <- data.frame()
  for(id in dtn){
    duplicates <- full.data[which(full.data$fixedID == id),]
    pc.fixedID <- rep(id, max(2, (length(duplicates[,1])-1)))
    pc.distance <- c()
    pc.tissue.concordance <- c()
    pc.library.concordance <- c()
    pc.lib.tissue.concordance.match <- c()
    for(d in 1:(length(duplicates[,1])-1)){
      pc.distance <- c(pc.distance, distance(x1 = duplicates$PC1[d],
                              x2 = duplicates$PC1[d+1],
                              y1 = duplicates$PC1[d],
                              y2 = duplicates$PC2[d+1]))
      if(duplicates$tissue[d] == duplicates$tissue[d+1]){
        pc.tissue.concordance <- c(pc.tissue.concordance, TRUE)
      }else{pc.tissue.concordance <- c(pc.tissue.concordance, FALSE)}
      if(duplicates$libName[d] == duplicates$libName[d+1]){
        pc.library.concordance <- c(pc.library.concordance, TRUE)
      }else{pc.library.concordance <- c(pc.library.concordance, FALSE)}
      if(tail(pc.tissue.concordance, n=1) == tail(pc.library.concordance, n=1)){
        pc.lib.tissue.concordance.match <- c(pc.lib.tissue.concordance.match, TRUE)
      }else{pc.lib.tissue.concordance.match <- c(pc.lib.tissue.concordance.match, FALSE)}
    }
    duplicate.data <- cbind(pc.fixedID, pc.distance, 
                            pc.tissue.concordance, pc.library.concordance,
                            pc.lib.tissue.concordance.match)
    reps.by.tissue.lib <- rbind(reps.by.tissue.lib, duplicate.data)
  }
  return(reps.by.tissue.lib)
}

replicate.concordance.data <- replicates_by_tissue_library(full.data.colors.85, NULL)
#todo -- is distance impacted by library or tissue concordance?

replicate.concordance.data$pc.distance <- as.numeric(as.character(replicate.concordance.data$pc.distance))
concordant.tissue <- replicate.concordance.data[which(replicate.concordance.data$pc.tissue.concordance == TRUE),]
disconcordant.tissue <- replicate.concordance.data[which(replicate.concordance.data$pc.tissue.concordance == FALSE),]
library(scales)
png(file='~/Desktop/Flinders_RADseq/TechReps_byTissue.png',
    height=12, width=15, units='cm', res=300)
par(mar=c(5,4.2,4,2))
hist((concordant.tissue$pc.distance), breaks = 12, probability = T, 
     col=alpha('blue', 0.5), border=alpha('blue', 0.01),
     main='Euclidian Distance between PC1 and PC2 for\nTechnical Replicates by Tissue',
     xlab='Distance', las=1, xlim=c(0,3),
     cex.axis=1.2, cex.lab=1.2, cex.main=1.2)
hist((disconcordant.tissue$pc.distance), breaks = 12, add=T, probability = T,
     col=alpha('red', 0.5), border=alpha('red', 0.01))
legend('topright', legend=c('Same Tissue', 'Different Tissue'), 
       fill=c(alpha('blue', 0.5), alpha('red', 0.5)), bty='n',
       border=c(alpha('blue', 0.01), alpha('red', 0.01)), cex=1.2)
dev.off()
```

How many sites vary between technical replicates?
```{r}
genotypes85<-read.table('~/Desktop/Flinders_RADseq/minDP10_minGQ25_missing85/dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing85_firstSNP_012.012')
names85<-read.table('~/Desktop/Flinders_RADseq/minDP10_minGQ25_missing85/dbr_filtered_all_samples_mapped_to_legs_larvae_pseudoref_genotypes_minDP10_minGQ25_missing85_firstSNP_012.012.indv')

hamming.distance <- function(seq1, seq2){
  if(length(seq1) != length(seq2)){
    return('Sequence lengths mismatch')
  }else{
    distance = 0
    for(i in 1:length(seq1)){
      if(seq1[i] == -1 | seq2[i] == -1){
        distance = distance + 2
      }else{distance = distance + abs(seq1[i]-seq2[i])}
      #0-2 or 2-0 = 2; 2-1 or 1-2 = 1; 0-0, 1-1, 2-2 = 0
    }
    return(as.numeric(distance))
  }
}

process.012 <- function(file.012, mat.size=NA){
  # fill in lower triangle of matrix
  # diagonal represents number of missing data points
  if(is.na(mat.size)){
    mat.size <- dim(file.012)[1]
  }
  dist.mat <- matrix(data=NA, nrow = as.numeric(mat.size), ncol = as.numeric(mat.size))
  dist.mat[mat.size, mat.size] <- hamming.distance(file.012[mat.size,],
                                                   file.012[mat.size,])
  for(c in 1:(mat.size-1)){
    dist.mat[c,c] <- hamming.distance(file.012[c,], file.012[c,])
    for(r in (c+1):mat.size){
      #fill upper triangle of distance matrix
      dist.mat[r,c] <- hamming.distance(file.012[r,], file.012[c,])
    }
  }
  return(dist.mat)
}
hamming.distance(genotypes85[1,], genotypes85[1,])
mat.85 <- process.012(genotypes85)
write.table(mat.85, 'distance_matrix_85missing.txt')

df.85 <- data.frame(mat.85)
df.85.named <- cbind(names85, df.85)
library(stringr)

find.replicates <- function(sample.names){
  dt <- table(sample.names)
  dtn <- names(dt[which(dt > 1)])  
  return(dtn)
}

fix.names <- function(distance.matrix){
  fixed.name <- c()
  for(n in 1:length(distance.matrix[,1])){
    #fn.tmp <- str_match(distance.matrix[n,1], "_[0-9]+_")
    fn.tmp <- substring(as.character(distance.matrix[n,1]), 133, 137)
    fn.tmp2 <- str_match(fn.tmp, ".*_")
    fn.tmp3 <- substring(fn.tmp2, 1, nchar(fn.tmp2)-1)
    fn.correct <- gsub("I", "1", fn.tmp3)
    fixed.name <- c(fixed.name, fn.correct)
  }
  distance.matrix[,1] <- fixed.name
  #return(fixed.name)
  return(distance.matrix)
}

df.85.fixed.names <- fix.names(df.85.named)
partial.matrix <- df.85.fixed.names[1:6, 1:6]

replicate.idx <- function(distance.matrix){
  # assume the matrix is sorted
  dups <- find.replicates(distance.matrix[,1])
  #dups <- 202
  ugh <- c()
  sample.id <- c()
  for(d in dups){
    sub.ugh <- c()
    indices <- rownames(distance.matrix[which(distance.matrix[,1] == d),])
    min.i <- as.numeric(as.character(min(indices)))
    max.i <- as.numeric(as.character(max(indices)))
    dup.sub <- distance.matrix[min.i:max.i,
               (min.i+1):(max.i+1)]
    dup.sub.lt <- lower.tri(dup.sub)
    for(i in 1:dim(dup.sub.lt)[1]){
      for(j in 1:dim(dup.sub.lt)[1]){
        if(dup.sub.lt[i,j]){
          sub.ugh <- c(sub.ugh, dup.sub[i,j])
        }
      }
    }
    ugh <- c(ugh, sub.ugh)
    sample.id <- c(sample.id, rep(d, length(sub.ugh)))
  }
  final <- cbind(sample.id, ugh)
  return(final)
}

duplicate.dist <- replicate.idx(df.85.fixed.names)

duplicate.dist <- data.frame(duplicate.dist)
names(duplicate.dist) <- c('sampleID', 'dist')
duplicate.dist$dist <- as.numeric(as.character(duplicate.dist$dist))
hist(duplicate.dist$dist)
boxplot(dist~sampleID, data=duplicate.dist, las=2, cex.axis = 0.6)
```
What is the null expectation for distance between two points in the space defined by PC1 and PC2?
```{r pca_null_distance}

```
### Old stuff to update for new data -- haven't gotten to this point yet...
```{r findClusters}
tick.c <- find.clusters(tick)
tick.dapc <- dapc(tick, pop = tick.c$grp)
scatter(tick.dapc)
scatter.colors = seasun(length(levels(tick.c$grp)))
plot(pca1$scores[,1], pca1$scores[,2], col = scatter.colors[tick.c$grp], pch=16)
hist(mat.85)
```

```{r assignmentProb}
compoplot(tick.dapc)
assignplot(tick.dapc)

posterior.cluster<-data.frame(tick.dapc$posterior) # extract posterior group assignments
posterior.cluster.named<-data.frame(cbind(row.names(posterior.cluster), posterior.cluster[,1], posterior.cluster[,2], posterior.cluster[,3], posterior.cluster[,4], posterior.cluster[,5])) # tweak the data structure to add sample names as a separate column (enables subsequent merging with data on collection site)
names(posterior.cluster.named)<-c('TID', 'clust.1', 'clust.2', 'clust.3', 'clust.4', 'clust.5')
posterior.cluster.named.spatial<-merge(posterior.cluster.named, metaData, by.x='TID', by.y='reconstructedID') # add site data to posteriors data frame by merging on the sample names

posterior.cluster.named.spatial<-posterior.cluster.named.spatial[order(posterior.cluster.named.spatial$'5'),]

short.pcns <- posterior.cluster.named.spatial[,2:6]
row.names(short.pcns)<-row.names(posterior.cluster.named.spatial)
ready.to.sort.pcns <- cbind(short.pcns, max.col(short.pcns))
sorted.pcns <- ready.to.sort.pcns[order(ready.to.sort.pcns$max.col),]
scatter.colors = seasun(length(levels(tick.c$grp)))
png(file='~/Desktop/Flinders_RADseq/AssignmentProbability.png', height=25, width=40, res=300, units='cm')
par(mar=c(6,4,2,8), xpd=TRUE)
barplot(t(sorted.pcns[1:5]), col=scatter.colors, axes=FALSE, ylab='Probability of Membership',
        xlim=c(0, length(posterior.cluster.named.spatial[,1])), 
        space=0, col.axis='white') #for some reason #s still show on x-axis... make their color white)
mtext(side=1, las=2, at=seq(0.5, length(sorted.pcns[,1])-0.5), cex=0.5, col=scatter.colors[sorted.pcns$max],
      text=row.names(sorted.pcns))
axis(side=2, las=1)
legend(x=375, y=0.55, legend=c('1', '2', '3', '4', '5'), fill=scatter.colors, title='Cluster \nAssignment', bty="n")
dev.off()
```

```{r consistency}
sorted.by.name <- posterior.cluster.named.spatial[order(posterior.cluster.named.spatial$fixedID),]
sorted.by.name.max <- cbind(sorted.by.name, max.col(sorted.by.name[2:6]))
do.duplicates.match <- cbind(sorted.by.name.max$fixedID, sorted.by.name.max$max)
counts<-table(do.duplicates.match[,1])
dups <- counts[which(counts > 1)]
n.unique <- c()
for(n in 1:length(names(dups))){
  subset.duplicates <- do.duplicates.match[which(do.duplicates.match[,1]==names(dups)[n]),]
  n.unique <- c(n.unique, length(unique(subset.duplicates[,2])))
}
dups.counts.accuracy <- rbind(dups, n.unique)
t.dca <- t(dups.counts.accuracy)
consistency <- abs(1-t.dca[,1]/t.dca[,2])
par(mar=c(4,4,2,2))
hist(consistency)

# fix the ID column...
metaData$fixedID <- gsub('I', '1', metaData$ID)
qualTime <- c()
for(id in 1:length(metaData[,1])){
  row = metaData[id,]
  print(row$fixedID)
  if(row$tissue == 'larvae'){
    qualTime <- c(qualTime, 'new')
  }else if(grepl('b', row$fixedID)){
      qualTime <- c(qualTime, 'new')
  }else if(as.numeric(as.character(row$fixedID)) < 381){
      qualTime <- c(qualTime, 'old')
  }else(qualTime <- c(qualTime, 'new'))
}
metaData$qualTime <- qualTime

time <- c()
for(t in 1:length(row.names(pca1$scores))){
  m <- metaData[which(as.character(metaData$reconstructedID) == as.character(row.names(pca1$scores)[t])),]
  #print(m)
  time <- c(time, as.character(m$qualTime))
}

time.colors <- c('#74c476','#005a32') #idx 1 = new, idx 2 = old
png('~/Desktop/Flinders_RADseq/PCA_colored_by_collection_time.png', width=15, height=15, res=300, unit='cm')
plot(pca1$scores[,1], pca1$scores[,2], col = time.colors[as.factor(time)], pch=16,
     xlab='PCA1', ylab='PCA2', las=1)
legend(x='topright', legend=c('2015', '2010'), col=time.colors, pch=16, bty='n')
dev.off()

# does dapc pick up library, tissue type, or collection time?

lib.dapc <- dapc(tick, pop = as.factor(lib))
scatter(lib.dapc)

tissue.dapc <- dapc(tick, pop = as.factor(tissue))
scatter(tissue.dapc)

time.dapc <- dapc(tick, pop = as.factor(time))
scatter(time.dapc)
```

Time of collection (2010 vs. 2015) seems to matter a lot -- not surprisingly.
We need to analyze the new and old ticks separately. We also have to look at geography.

```{r geography}
spatial.data <- read.csv('~/Dropbox/TrugosaTicks/DataSheets/MASTER_DATA_SHEET_AllSamples_Tab.csv')

# use metaData and spatial.data and pca1 to combine pcs w/GPS coords
eastings <- c()
northings <- c()
simpleID <- c()
coll.date <- c()
for(t in 1:length(row.names(pca1$scores))){
  print(t)
  #print(as.character(row.names(pca1$scores)[t]))
  # look up the tick entry by full barcoded name
  m <- metaData[which(as.character(metaData$reconstructedID) == as.character(row.names(pca1$scores)[t])),]
  #print(m)
  # look up the spatial data using the simple tick ID number
  s <- spatial.data[which(as.character(m$fixedID) == as.character(spatial.data$TickNumber)),]
  print(factor(s[,1]))
  if(length(s[,1])>0){
    print(factor(s[,1]))
    eastings <- c(eastings, as.integer(as.character(s$Easting)))
    northings <- c(northings, as.integer(as.character(s$Northing)))
    simpleID <- c(simpleID, as.character(s$TickNumber))
    coll.date <- c(coll.date, as.character(s$CollectionDate))
  }else{
    eastings <- c(eastings, NA)
    northings <- c(northings, NA)
    simpleID <- c(simpleID, NA)
    } # apparently this else block does not get evaluated...
}

pca.spatial <- data.frame(cbind(simpleID,
                                coll.date,
                                scatter.colors[tick.c$grp], 
                                eastings, northings))
                                # tick.c$grp has same order as pca1$scores so the colors will be correct

with.data <- pca.spatial[complete.cases(pca.spatial),]
sunflowerplot(as.integer(as.character(with.data$eastings)), 
     as.integer(as.character(with.data$northings)), 
     seg.col=as.character(with.data[,3]), pch=16)

with.data$uniqueCoord <- paste(with.data$eastings, with.data$northings, sep='_')
color.coords <- cbind(as.character(with.data[,3]), with.data$uniqueCoord)
summary.colors <- table(color.coords[,1], color.coords[,2])

library(plotrix)
#floating.pie(400000, 6180000,
#             x=summary.colors[,1], radius = 5000)

column.names <- dimnames(summary.colors)[[2]]

plot(x=NA, y=NA, xlim=c(min(as.integer(as.character(with.data$eastings)))-1000,
                max(as.integer(as.character(with.data$eastings)))+1000),
         ylim=c(min(as.integer(as.character(with.data$northings)))-1000,
                max(as.integer(as.character(with.data$northings)))+1000))
for(column in 1:length(column.names)){
  coords <- strsplit(column.names[column], split = '_')
  u.coords <- unlist(coords)
  floating.pie(as.integer(as.character(u.coords[1])), 
               as.integer(as.character(u.coords[2])), 
               x=summary.colors[,column], 
               col=row.names(summary.colors),
               radius = 3000)
}


# need to collapse into grid... 1ha grid:
with.data$eastings<-as.numeric(as.character(with.data$eastings))
with.data$northings<-as.numeric(as.character(with.data$northings))

# make 10km x 10km grid
# extend the range below the minimum coordinates and offset by 5 to prevent exact matches
eastingSeq = seq(min(with.data$eastings)-10000, 
                 max(with.data$eastings)+10000, 10000)
northingSeq = seq(min(with.data$northings)-10000, 
                  max(with.data$northings)+10000, 10000)
assignedGrid = c()
# there has got to be a better way of assigning points to grid cells than this...
for(i in 1:(length(eastingSeq)-1)){
  for(j in 1:(length(northingSeq)-1)){
    inE <- with.data[which(eastingSeq[i] < with.data$eastings
                                  & with.data$eastings <= eastingSeq[i+1]),]
    #print(inE)
    if(length(inE[,1])>0){
      inN <- inE[which(northingSeq[j] < inE$northings 
                       & inE$northings <= northingSeq[j+1]),]
      if(length(inN[,1])>0){
        #print(inN)
        inN$gridE <- rep(i, length(inN[,1]))
        inN$gridN <- rep(j, length(inN[,1]))
        inN$centerEasting = rep((eastingSeq[i]+eastingSeq[i+1])/2, length(inN[,1]))
        inN$centerNorthing = rep((northingSeq[j]+northingSeq[j+1])/2, length(inN[,1]))
        print(paste(i, j, sep='-'))
        print(inN)
        assignedGrid = rbind(assignedGrid, inN)
      }
    }
  }
}

assignedGrid$uniqueCoord <- paste(assignedGrid$eastings, assignedGrid$northings, sep='_')
assignedGrid$uniqueGrid <- paste(assignedGrid$gridE, assignedGrid$gridN, sep='_')
summary.colors.ha <- table(assignedGrid$V3, assignedGrid$uniqueGrid)

column.names.ha <- dimnames(summary.colors.ha)[[2]]

# organize collection time
proper.dates <- strsplit(as.character(assignedGrid$coll.date), split = '-')
yr <- c()
for(p in 1:length(proper.dates)){
  item <- proper.dates[[p]]
  if(length(item)==3){
    yr <- c(yr, proper.dates[[p]][3])
  }else{yr <- c(yr, 'missing')}
}
assignedGrid$coll.year <- yr
assignedGrid$uniqueCenter <- paste(assignedGrid$centerEasting, 
                                   assignedGrid$centerNorthing, sep = '_')
summary.colors.time <- table(assignedGrid$uniqueCenter, assignedGrid$coll.year)
time.colors <- c('white', 'black', 'gray') 
dimnames(summary.colors.time)[[2]] <- time.colors
rns <- dimnames(summary.colors.time)[[1]]

# make the plots
#nf <- layout(matrix(c(1,2),2,2,byrow=TRUE), widths=c(16,8), heights=c(10,10), TRUE)
#layout.show(nf)
png(file='~/Desktop/Flinders_RADseq/ClusterPieChart_CoarseMap.png',
    height=15, width=20, units='cm', res=300)
par(mar=c(5,5,2,2))
plot(x=NA, y=NA, xlim=c(min(as.integer(as.character(assignedGrid$centerEasting)))-1000,
                max(as.integer(as.character(assignedGrid$centerEasting)))+50000),
         ylim=c(min(as.integer(as.character(assignedGrid$centerNorthing)))-1000,
                max(as.integer(as.character(assignedGrid$centerNorthing)))+20000),
     axes=F, xlab='Relative Easting (km)', ylab='Relative Northing (km)')
for(row in 1:length(rns)){
  getC <- assignedGrid[which(assignedGrid$uniqueCenter == rns[row]),]
  east <- as.integer(as.character(getC$centerEasting[1]))
  north <- as.integer(as.character(getC$centerNorthing[1]))
  use.colors <- c()
  for(value in 1:length(summary.colors.time[row,])){
    if(summary.colors.time[row, value] != 0){
      use.colors <- c(use.colors, time.colors[value])
    }
  }
  floating.pie(east, north, x=summary.colors.time[row,], 
               col = use.colors, radius = 4500)
}
cluster.colors <- row.names(summary.colors.ha)
for(column in 1:length(column.names.ha)){
  getC <- assignedGrid[which(assignedGrid$uniqueGrid == column.names.ha[column]),]
  east <- as.integer(as.character(getC$centerEasting[1]))
  north <- as.integer(as.character(getC$centerNorthing[1]))
  use.colors <- c()
  for(value in 1:length(summary.colors.ha[,column])){
    if(summary.colors.ha[value, column] != 0){
      use.colors <- c(use.colors, cluster.colors[value])
    }
  }
  floating.pie(east, north, x=summary.colors.ha[,column], 
               col = use.colors, radius = 2800)
}
axis(side=1, at=eastingSeq, labels=(eastingSeq-min(eastingSeq))/1000)
axis(side=2, at=northingSeq, labels=(northingSeq-min(northingSeq))/1000, las=1)

text(x=max(as.integer(as.character(assignedGrid$centerEasting)))+15000,
     y=6250000, labels='Collection Year')
floating.pie(max(as.integer(as.character(assignedGrid$centerEasting)))+15000, 
                 6240000, x=1, col= 'black', radius = 4500)
text(x=max(as.integer(as.character(assignedGrid$centerEasting)))+25000,
     y=6240000, labels='2015', adj=0)
draw.circle(max(as.integer(as.character(assignedGrid$centerEasting)))+15000, 
                6240000, radius = 2800, col= 'white')
floating.pie(max(as.integer(as.character(assignedGrid$centerEasting)))+15000, 
                 6240000-10000, x=1, col= 'gray', radius = 4500)
text(x=max(as.integer(as.character(assignedGrid$centerEasting)))+25000,
     y=6240000-10000, labels='missing', adj=0)
draw.circle(max(as.integer(as.character(assignedGrid$centerEasting)))+15000, 
                6240000-10000, radius = 2800, col= 'white')
floating.pie(max(as.integer(as.character(assignedGrid$centerEasting)))+15000, 
                 6240000-20000, x=1, col= 'white', radius = 4500)
text(x=max(as.integer(as.character(assignedGrid$centerEasting)))+25000,
     y=6240000-20000, labels='2010', adj=0)
draw.circle(max(as.integer(as.character(assignedGrid$centerEasting)))+15000, 
                6240000-20000, radius = 2800, col= 'white', border='black')
# legend for inner pies (inferred cluster)
text(x=max(as.integer(as.character(assignedGrid$centerEasting)))+15000,
     y=6240000-30000, labels='Cluster')
floating.pie(max(as.integer(as.character(assignedGrid$centerEasting)))+15000, 
             6240000-40000, x=1, col= cluster.colors[1], radius = 2800)
text(x=max(as.integer(as.character(assignedGrid$centerEasting)))+25000,
     y=6240000-40000, labels='1', adj=0)
floating.pie(max(as.integer(as.character(assignedGrid$centerEasting)))+15000, 
             6240000-50000, x=1, col= cluster.colors[2], radius = 2800)
text(x=max(as.integer(as.character(assignedGrid$centerEasting)))+25000,
     y=6240000-50000, labels='2', adj=0)
floating.pie(max(as.integer(as.character(assignedGrid$centerEasting)))+15000, 
             6240000-60000, x=1, col= cluster.colors[3], radius = 2800)
text(x=max(as.integer(as.character(assignedGrid$centerEasting)))+25000,
     y=6240000-60000, labels='3', adj=0)
floating.pie(max(as.integer(as.character(assignedGrid$centerEasting)))+15000, 
             6240000-70000, x=1, col= cluster.colors[4], radius = 2800)
text(x=max(as.integer(as.character(assignedGrid$centerEasting)))+25000,
     y=6240000-70000, labels='4', adj=0)
floating.pie(max(as.integer(as.character(assignedGrid$centerEasting)))+15000, 
             6240000-80000, x=1, col= cluster.colors[5], radius = 2800)
text(x=max(as.integer(as.character(assignedGrid$centerEasting)))+25000,
     y=6240000-80000, labels='5', adj=0)
dev.off()

# helpful: http://rgraphics.limnology.wisc.edu/rmargins_layout.php
require(maptools)
require(ggplot2)

murray <- readShapePoly('~/Downloads/WATER_RiverMurray_shp/WATER_RiverMurray.shp')
mf <- fortify(murray)
proj4string(mf)<-CRS("+proj=longlat +datum=WGS84")
mf_UTM<-spTransform(mf, CRS("+proj=utm +zone=54 +south +ellps=WGS84"))

points(mf$long, mf$lat)
```